### PR TITLE
feat: add login command

### PR DESCRIPTION
**Pull request by JivusAyrus**
- [ ] Tests or benchmark included
- [ ] Documentation is changed or added on [https://app.gitbook.com/](https://app.gitbook.com/)
- [x] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)

FIXES ENG-4075

### CODE REVIEW

File: cli/src/commands/auth/commands/login.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      }
+    }
+
+    const resp = await performDeviceAuth();
+    if (!resp.success) {
+      program.error('Could not perform authentication. Please try again');
+    }
+    console.log('Code: %s\n', resp.response.deviceCode);
+    console.log(
+      'If your browser does not automatically open, use this URL and enter the Code to verify the login: %s\n',
```
Comment (StarpTech) [inline cli/src/commands/auth/commands/login.ts:31]: :+1:

File: cli/src/commands/auth/commands/login.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      const loginData = JSON.parse(JSON.stringify(data));
+      if (loginData && loginData?.expiresAt && new Date(loginData.expiresAt) > new Date()) {
+        console.log(pc.green('You are already logged in.'));
+        process.exit(0);
+      }
+    }
+
+    const resp = await performDeviceAuth();
+    if (!resp.success) {
+      program.error('Could not perform authentication. Please try again');
```
Comment (StarpTech) [inline cli/src/commands/auth/commands/login.ts:27]: Can we show some debug information? or is it logged before?

File: controlplane/src/core/services/AccessTokenAuthenticator.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  public async authenticate(accessToken: string): Promise<AccessTokenAuthContext> {
+    const userInfoData = await this.authUtils.getUserInfo(accessToken);
+
+    // TODO will have to change when wesupport multiple orgs
+    const orgSlug = userInfoData.groups[0].split('/')[1];
+
+    const organization = await this.orgRepo.bySlug(orgSlug);
+
+    if (!organization || !organization?.id) {
+      throw new Error('Organization deosnt exist');
```
Comment (StarpTech) [inline controlplane/src/core/services/AccessTokenAuthenticator.ts:26]: typo and throw Unauthorized error here so we don't log this as an error.

File: cli/src/core/config.ts [hunk 4]
```diff
@@ -1,14 +1,33 @@
...
+import path from 'node:path';
+import yaml from 'js-yaml';
 
 const info = JSON.parse(
   await readFile(new URL('../../package.json', import.meta.url), {
     encoding: 'utf8',
   }),
 );
 
+export const configDir = path.join(os.homedir(), '.cosmo');
```
Comment (StarpTech) [inline cli/src/core/config.ts:13]: I'm not sure if this is the right approach for all systems. I'd take inspiration from this package https://github.com/sindresorhus/env-paths

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: cli/src/commands/auth/commands/login.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    if (existsSync(configFile)) {
+      const data = yaml.load(readFileSync(configFile, 'utf8'));
+      const loginData = JSON.parse(JSON.stringify(data));
+      if (loginData && loginData?.expiresAt && new Date(loginData.expiresAt) > new Date()) {
+        console.log(pc.green('You are already logged in.'));
+        process.exit(0);
+      }
+    }
+
+    const resp = await performDeviceAuth({ cliClientId: 'cosmo-cli' });
```
[the remark this answers is not in the document]
Comment (JivusAyrus) [reply cli/src/commands/auth/commands/login.ts]: Done

File: controlplane/src/core/services/AccessTokenAuthenticator.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+  /**
+   * Authenticates the user with the given access token. Returns the user's organization ID and user's ID.
+   *
+   * @param accessToken
+   */
+  public async authenticate(accessToken: string): Promise<AccessTokenAuthContext> {
+    const userInfoData = await this.authUtils.getUserInfo(accessToken);
+
+    // TODO will have to change when wesupport multiple orgs
```
Comment (StarpTech) [inline controlplane/src/core/services/AccessTokenAuthenticator.ts:20]: typo

File: cli/src/commands/auth/commands/login.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+import { configDir, configFile } from '../../../core/config.js';
+import { BaseCommandOptions } from '../../../core/types/types.js';
+import { performDeviceAuth, startPollingForAccessToken } from '../utils.js';
+
+export default (opts: BaseCommandOptions) => {
+  const loginCommand = new Command('login');
+  loginCommand.description('Login a user.');
+
+  loginCommand.action(async () => {
+    if (existsSync(configFile)) {
```
[the remark this answers is not in the document]
Comment (JivusAyrus) [reply cli/src/commands/auth/commands/login.ts:16]: which helper functions?

File: controlplane/src/core/services/AccessTokenAuthenticator.ts (hunk 6 above)
Comment (JivusAyrus) [reply controlplane/src/core/services/AccessTokenAuthenticator.ts]: Done

File: cli/src/core/config.ts (hunk 4 above)
Comment (JivusAyrus) [reply cli/src/core/config.ts]: Done

File: controlplane/src/core/services/AccessTokenAuthenticator.ts (hunk 3 above)
Comment (JivusAyrus) [reply controlplane/src/core/services/AccessTokenAuthenticator.ts]: Done

File: cli/src/commands/auth/commands/login.ts [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+import yaml from 'js-yaml';
+import open from 'open';
+import pc from 'picocolors';
+import { configDir, configFile } from '../../../core/config.js';
+import { BaseCommandOptions } from '../../../core/types/types.js';
+import { performDeviceAuth, startPollingForAccessToken } from '../utils.js';
+
+export default (opts: BaseCommandOptions) => {
+  const loginCommand = new Command('login');
+  loginCommand.description('Login a user.');
```
[the remark this answers is not in the document]
Comment (JivusAyrus) [reply cli/src/commands/auth/commands/login.ts]: dont know what else to write, changed it to "Login as an user"

File: cli/src/commands/auth/commands/login.ts (hunk 2 above)
Comment (JivusAyrus) [reply cli/src/commands/auth/commands/login.ts:21]: i dont think there is something we could show

File: cli/src/commands/auth/commands/login.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (JivusAyrus) [reply cli/src/commands/auth/commands/login.ts]: like this https://github.com/wundergraph/cosmo/pull/95/commits/71f39736c3397a1703d8835418c122a25e533ad3#:~:text=export%20const%20checkConfigFile,%7D%3B ?

[the remark this answers is not in the document]
Comment (StarpTech) [reply cli/src/commands/auth/commands/login.ts]: Not exactly. A util function would be an implementation that is usable in multiple places. I had in mind e.g. a file called `user-config.ts` that provides load and `validateToken` functions but you can move it back if you think it's not worth it.

File: cli/src/commands/auth/commands/login.ts (hunk 2 above)
Comment (StarpTech) [reply cli/src/commands/auth/commands/login.ts:21]: ok

File: cli/src/commands/auth/commands/login.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (StarpTech) [reply cli/src/commands/auth/commands/login.ts]: > Login a user into the Cosmo platform. Supports browser-based authentication.

File: cli/src/commands/auth/commands/login.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (JivusAyrus) [reply cli/src/commands/auth/commands/login.ts]: Done

File: cli/src/commands/auth/commands/login.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (JivusAyrus) [reply cli/src/commands/auth/commands/login.ts]: Done

**Review by StarpTech — APPROVED** [review]
LGTM

**Pull request merged**